### PR TITLE
haskellPackages.cereal: fix for pre-GHC8

### PR DESCRIPTION
**Pull request by Profpatsch**
See commit message.
Should fix git-annex.

Next step would be to move all packages that deal with this kind of issue from 7.10 overrides to common.
Next step would be to make cabal2nix recognize conditionals.

cc @peti

### CODE REVIEW

File: pkgs/development/haskell-modules/configuration-lts.nix [hunk 1]
```diff
@@ -1783,7 +1783,6 @@ self: super: {
   "ceilometer-common" = dontDistribute super."ceilometer-common";
   "cellrenderer-cairo" = dontDistribute super."cellrenderer-cairo";
   "cerberus" = dontDistribute super."cerberus";
-  "cereal" = doDistribute super."cereal_0_5_2_0";
```
Comment (bennofs) [inline pkgs/development/haskell-modules/configuration-lts.nix]: Why did you remove this override? https://www.stackage.org/lts-6.9 looks like it still has that cereal version

Comment (Profpatsch) [inline pkgs/development/haskell-modules/configuration-lts.nix]: Oh, wasn’t aware that it was automatically generated. Adding it back in …

Comment (Profpatsch) [inline pkgs/development/haskell-modules/configuration-lts.nix]: Hm, that makes it fail again.

I wonder if I should add the same rule for 5.2 or just keep 5.3, since it works.

Comment (bennofs) [inline pkgs/development/haskell-modules/configuration-lts.nix]: I think you should add the same rule for 5.2 if 5.2 fails as well.

Comment (Profpatsch) [inline pkgs/development/haskell-modules/configuration-lts.nix]: Done.

File: pkgs/development/haskell-modules/configuration-common.nix [hunk 2]
```diff
@@ -2,7 +2,17 @@
 
 with import ./lib.nix { inherit pkgs; };
 
-self: super: {
+self: super:
+
+let
+  # apply an override to a package if it’s build with GHC <8.0
+  ifLT8 = override: drv:
```
Comment (peti) [inline pkgs/development/haskell-modules/configuration-common.nix]: This function should be in `lib.nix` and it should have a more descriptive name than `ifLT8`.

File: pkgs/development/haskell-modules/configuration-common.nix [hunk 3]
```diff
@@ -947,8 +957,11 @@ self: super: {
...
-  Glob = if pkgs.lib.versionAtLeast self.ghc.version "8.0"
-    then super.Glob
-    else addBuildDepend super.Glob self.semigroups;
+
+  # packages that depend on extra packages for GHC <8.0
+  Glob = ifLT8 (p: addBuildDepend p self.semigroups) super.Glob;
+  # cereal must have `fail` in pre-ghc-8.0.x versions
+  # also tests require bytestring>=0.10.8.1
+  cereal = ifLT8 (p: dontCheck (addBuildDepend p self.fail)) super.cereal;
+  cereal_0_5_2_0 = ifLT8 (p: dontCheck (addBuildDepend p self.fail)) super.cereal_0_5_2_0;
```
Comment (peti) [inline pkgs/development/haskell-modules/configuration-common.nix]: What I don't like about this approach is that we have a file `configuration-ghc-7.10.x.nix` that says "overrides applied for GHC 7.10.x are in here!", but in fact they are not. There is something to be said for duplicating those overrides in the various compiler configs, even if it is more verbose.

File: pkgs/development/haskell-modules/configuration-common.nix (hunk 2 above)
Comment (Profpatsch) [inline pkgs/development/haskell-modules/configuration-common.nix]: I didn’t want to expose that (yet), since it’s not completely clear if it’s a helpful library function (at least for me).

File: pkgs/development/haskell-modules/configuration-common.nix (hunk 3 above)
Comment (Profpatsch) [inline pkgs/development/haskell-modules/configuration-common.nix]: Would it make sense to have a new overwrite layer for these kind of overwrites? At least until we can generate it with cabal2nix? Something like `configuration-old-compilers.nix` maybe.

File: pkgs/development/haskell-modules/configuration-common.nix (hunk 2 above)
Comment (peti) [inline pkgs/development/haskell-modules/configuration-common.nix]: Yeah, but it's confusing for readers of the file to define a "local helper function" some 1000 lines of code apart from the place where it's actually used.

File: pkgs/development/haskell-modules/configuration-common.nix (hunk 3 above)
Comment (peti) [inline pkgs/development/haskell-modules/configuration-common.nix]: I feel like the current solution -- copy-pasting the override into those config files that need it -- works sufficiently well and I don't see much of an incentive to add another config file right now. Besides, it's not uncommon for those overrides to require even more extra build inputs the older the compiler gets, i.e. it's not always easy to define one override that works everywhere.

Comment (Profpatsch) [inline pkgs/development/haskell-modules/configuration-common.nix]: > Besides, it's not uncommon for those overrides to require even more extra build inputs the older the compiler gets

In that case the function can be applied multiple times with different compiler versions.
So should I copy and paste that into every 7.x configuration file?

Comment (peti) [inline pkgs/development/haskell-modules/configuration-common.nix]: > So should I copy and paste that into every 7.x configuration file?

I would prefer that.

**Comment by Profpatsch** [comment]
@peti I copied it to 7.10, so it should now work for lts as well.
